gusto/equations/shallow_water_equations.py

```diff
@@ -272,7 +232,245 @@ def __init__(self, domain, parameters, fexpr=None, bexpr=None,
+            lambda t: t.has_label(pressure_gradient),
+            drop)
+
+        # add (moist) thermal source terms not involving topography -
+        # label these as the equivalent pressure gradient term
+        if equivalent_buoyancy:
+            q0 = parameters.q0
+            nu = parameters.nu
+            beta2 = parameters.beta2
+            qsat_expr = compute_saturation(q0, nu, H, g, D, b_name, topog)
+            qv = conditional(qt < qsat_expr, qt, qsat_expr)
+            source_form = pressure_gradient(subject(prognostic(
+                -D * div(b*w) * dx - 0.5 * b * div(D*w) * dx
+                + jump(b*w, n) * avg(D) * dS + 0.5 * jump(D*w, n) * avg(b) * dS
+                - beta2 * D * div(qv*w)*dx - 0.5 * beta2 * qv * div(D*w) * dx
+                + beta2 * jump(qv*w, n) * avg(D) * dS
+                + 0.5 * beta2 * jump(D*w, n) * avg(qv) * dS,
+                'u'), self.X))
+        else:
+            source_form = pressure_gradient(
+                subject(prognostic(-D * div(b*w) * dx
+                                   + jump(b*w, n) * avg(D) * dS
+                                   - 0.5 * b * div(D*w) * dx
+                                   + 0.5 * jump(D*w, n) * avg(b) * dS,
+                                   'u'), self.X))
+        residual += source_form
+
+        # -------------------------------------------------------------------- #
+        # add transport terms:
+        # -------------------------------------------------------------------- #
+        b_adv = prognostic(advection_form(gamma, b, u), b_name)
+        residual += subject(b_adv, self.X)
+
+        if equivalent_buoyancy:
+            qt_adv = prognostic(advection_form(gamma_qt, qt, u), "q_t")
+            residual += subject(qt_adv, self.X)
+
+        # -------------------------------------------------------------------- #
+        # add topography terms:
+        # -------------------------------------------------------------------- #
+        if bexpr is not None:
+            if equivalent_buoyancy:
+                topography_form = subject(prognostic(
+                    -topog * div(b*w) * dx
+                    + jump(b*w, n) * avg(topog) * dS
+                    - beta2 * topog * div(qv*w) * dx
+                    + beta2 * jump(qv*w, n) * avg(topog) * dS,
+                    'u'), self.X)
+            else:
+                topography_form = subject(prognostic(
+                    -topog * div(b*w) * dx
+                    + jump(b*w, n) * avg(topog) * dS,
+                    'u'), self.X)
+            residual += topography_form
+
+        # -------------------------------------------------------------------- #
+        # Linearise equations
+        # -------------------------------------------------------------------- #
+        # Add linearisations to equations
+        self.residual = self.generate_linear_terms(residual, self.linearisation_map)
+
+    def setup(self):
+        self.field_names = ['u', 'D']
```

Comment:
Do we need this `setup` function? Could we just set `self.field_names` and `self.space_names` in the `__init__` method?

Reply:
ugh, yeah, I don't like this either and there might be a better way of doing it... The problem is that the mixed function space is set up in the `PrognosticEquationSet` `__init__` method which gets  passed `field_names` and `spaces` from the `ShallowWaterEquations` `__init__` method... so if we set those things in the `__init__` method of `ThermalShallowWaterEquations` then they get overwritten in `ShallowWaterEquations`.